File: src/main/java/wraith/fabricaeexnihilo/modules/crucibles/CrucibleBlockEntity.java
package wraith.fabricaeexnihilo.modules.crucibles;

import net.fabricmc.fabric.api.object.builder.v1.block.entity.FabricBlockEntityTypeBuilder;
import net.fabricmc.fabric.api.transfer.v1.context.ContainerItemContext;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidConstants;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidStorage;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.fabricmc.fabric.api.transfer.v1.item.ItemStorage;
import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant;
import net.fabricmc.fabric.api.transfer.v1.storage.Storage;
import net.fabricmc.fabric.api.transfer.v1.storage.StorageUtil;
import net.fabricmc.fabric.api.transfer.v1.storage.base.ExtractionOnlyStorage;
import net.fabricmc.fabric.api.transfer.v1.storage.base.InsertionOnlyStorage;
import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleSlotStorage;
import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
import net.fabricmc.fabric.api.transfer.v1.transaction.base.SnapshotParticipant;
import net.minecraft.block.BlockState;
import net.minecraft.block.FluidBlock;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.jetbrains.annotations.Nullable;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModBlocks;
import wraith.fabricaeexnihilo.modules.barrels.BarrelBlock;
import wraith.fabricaeexnihilo.modules.base.BaseBlockEntity;
import wraith.fabricaeexnihilo.modules.base.EnchantableBlockEntity;
import wraith.fabricaeexnihilo.modules.base.EnchantmentContainer;
import wraith.fabricaeexnihilo.recipe.crucible.CrucibleHeatRecipe;
import wraith.fabricaeexnihilo.recipe.crucible.CrucibleRecipe;
import wraith.fabricaeexnihilo.util.CodecUtils;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

@SuppressWarnings("UnstableApiUsage")
public class CrucibleBlockEntity extends BaseBlockEntity implements EnchantableBlockEntity {

    public static final Identifier BLOCK_ENTITY_ID = id("crucible");

    public static final BlockEntityType<CrucibleBlockEntity> TYPE = FabricBlockEntityTypeBuilder.create(
        CrucibleBlockEntity::new,
        ModBlocks.CRUCIBLES.values().toArray(new CrucibleBlock[0])
    ).build(null);

    static {
        ItemStorage.SIDED.registerForBlockEntity((crucible, direction) -> crucible.itemStorage, TYPE);
        FluidStorage.SIDED.registerForBlockEntity((crucible, direction) -> crucible.fluidStorage, TYPE);
    }

    /**
     * Enchantments
     */
    private final EnchantmentContainer enchantments = new EnchantmentContainer();
    private final Storage<FluidVariant> fluidStorage = new CrucibleFluidStorage();
    private final Storage<ItemVariant> itemStorage = new CrucibleItemStorage();
    private long contained = 0;
    private FluidVariant fluid = FluidVariant.blank();
    private int heat = 0;
    private long queued = 0;
    private ItemStack renderStack = ItemStack.EMPTY;
    private int tickCounter;
    
    public CrucibleBlockEntity(BlockPos pos, BlockState state) {
        super(TYPE, pos, state);
        tickCounter = world == null
            ? FabricaeExNihilo.CONFIG.modules.crucibles.tickRate
            : world.random.nextInt(FabricaeExNihilo.CONFIG.modules.crucibles.tickRate);
    }

    @SuppressWarnings("unused") // lambda stuff
    public static void ticker(World world, BlockPos blockPos, BlockState blockState, CrucibleBlockEntity crucibleEntity) {
        crucibleEntity.tick();
    }

    public ActionResult activate(@Nullable PlayerEntity player, @Nullable Hand hand) {
        if (world == null || player == null) {
            return ActionResult.PASS;
        }
        var held = player.getStackInHand(hand == null ? player.getActiveHand() : hand);
        if (held == null || held.isEmpty()) {
            return ActionResult.PASS;
        }

        var bucketFluidStorage = FluidStorage.ITEM.find(held, ContainerItemContext.ofPlayerHand(player, hand));
        if (bucketFluidStorage != null) {
            var amount = StorageUtil.move(fluidStorage, bucketFluidStorage, fluid -> true, Long.MAX_VALUE, null);
            if (amount > 0) {
                markDirty();
                return ActionResult.SUCCESS;
            }
        }

        try (var t = Transaction.openOuter()) {
            var amount = itemStorage.insert(ItemVariant.of(held), 1, t);
            if (amount > 0) {
                t.commit();
                if (!player.isCreative()) {
                    held.decrement((int) amount);
                }
                markDirty();
                return ActionResult.SUCCESS;
            }
        }

        return ActionResult.PASS;
    }

    public long getContained() {
        return contained;
    }

    private int getEfficiencyMultiplier() {
        return 1 + enchantments.getEnchantmentLevel(Enchantments.EFFICIENCY);
    }

    @Override
    public EnchantmentContainer getEnchantmentContainer() {
        return enchantments;
    }

    public EnchantmentContainer getEnchantments() {
        return enchantments;
    }

    private int getFireAspectAdder() {
        return enchantments.getEnchantmentLevel(Enchantments.FIRE_ASPECT);
    }

    public FluidVariant getFluid() {
        return fluid;
    }

    public long getMaxCapacity() {
        return FluidConstants.BUCKET * (isFireproof() ? FabricaeExNihilo.CONFIG.modules.crucibles.stoneVolume : FabricaeExNihilo.CONFIG.modules.crucibles.woodVolume);
    }

    public long getProcessingSpeed() {
        return getEfficiencyMultiplier() * (isFireproof() ? heat * FabricaeExNihilo.CONFIG.modules.crucibles.baseProcessRate : FabricaeExNihilo.CONFIG.modules.crucibles.woodenProcessingRate);
    }

    public long getQueued() {
        return queued;
    }

    public ItemStack getRenderStack() {
        return renderStack;
    }
    
    public boolean isFireproof() {
        return getCachedState().getBlock() instanceof CrucibleBlock crucible && crucible.isFireproof();
    }

    @Override
    public void readNbt(NbtCompound nbt) {
        super.readNbt(nbt);
        if (nbt == null) {
            FabricaeExNihilo.LOGGER.warn("A crucible at " + pos + " is missing data.");
            return;
        }
        readNbtWithoutWorldInfo(nbt);
    }

    private void readNbtWithoutWorldInfo(NbtCompound nbt) {
        renderStack = ItemStack.fromNbt(nbt.getCompound("render"));
        fluid = CodecUtils.fromNbt(CodecUtils.FLUID_VARIANT, nbt.get("fluid"));
        contained = nbt.getLong("contained");
        queued = nbt.getLong("queued");
        heat = nbt.getInt("heat");
        // Why? It always exists...
        if (nbt.contains("enchantments")) {
            var readEnchantments = new EnchantmentContainer();
            readEnchantments.readNbt(nbt.getCompound("enchantments"));
            enchantments.setAllEnchantments(readEnchantments);
        }
    }

    public void tick() {
        if (queued == 0 || contained > getMaxCapacity() || (heat <= 0 && isFireproof())) {
            return;
        }
        if (--tickCounter <= 0) {
            var amount = Math.min(queued, getProcessingSpeed());
            contained += amount;
            queued -= amount;
            markDirty();
            tickCounter = FabricaeExNihilo.CONFIG.modules.crucibles.tickRate;
        }
    }

    public void updateHeat() {
        if (world == null) {
            return;
        }
        var oldHeat = heat;
        var state = world.getBlockState(pos.down());
        var block = state.getBlock();
        heat = CrucibleHeatRecipe.find(block, world).map(CrucibleHeatRecipe::getHeat).orElse(0);
        if (block instanceof FluidBlock) {
            heat *= state.getFluidState().getHeight();
        }
        heat += getFireAspectAdder();
        if (heat != oldHeat) {
            markDirty();
        }
    }

    /**
     * NBT Serialization section
     */

    @Override
    public void writeNbt(NbtCompound nbt) {
        super.writeNbt(nbt);
        writeNbtWithoutWorldInfo(nbt);
    }

    private void writeNbtWithoutWorldInfo(NbtCompound nbt) {
        nbt.put("render", renderStack.writeNbt(new NbtCompound()));
        nbt.put("fluid", CodecUtils.toNbt(CodecUtils.FLUID_VARIANT, fluid));
        nbt.putLong("contained", contained);
        nbt.putLong("queued", queued);
        nbt.putInt("heat", heat);
        nbt.put("enchantments", enchantments.writeNbt());
    }

    private record CrucibleSnapshot(long contained, long queued, FluidVariant fluid, ItemStack renderStack) {
    }

    private class CrucibleFluidStorage extends SnapshotParticipant<CrucibleSnapshot> implements SingleSlotStorage<FluidVariant>, ExtractionOnlyStorage<FluidVariant> {

        @Override
        protected CrucibleSnapshot createSnapshot() {
            return new CrucibleSnapshot(contained, queued, fluid, renderStack);
        }

        @Override
        public long extract(FluidVariant resource, long maxAmount, TransactionContext transaction) {
            if (!resource.equals(fluid))
                return 0;

            var amount = Math.min(maxAmount, contained);
            updateSnapshots(transaction);
            contained -= amount;
            return amount;
        }

        @Override
        public long getAmount() {
            return contained;
        }

        @Override
        public long getCapacity() {
            return getMaxCapacity();
        }

        @Override
        public FluidVariant getResource() {
            return fluid;
        }

        @Override
        public boolean isResourceBlank() {
            return fluid.isBlank();
        }

        @Override
        protected void readSnapshot(CrucibleSnapshot snapshot) {
            contained = snapshot.contained;
            queued = snapshot.queued;
            fluid = snapshot.fluid;
            renderStack = snapshot.renderStack.copy();
        }
    }

    private class CrucibleItemStorage extends SnapshotParticipant<CrucibleSnapshot> implements InsertionOnlyStorage<ItemVariant>, SingleSlotStorage<ItemVariant> {

        @Override
        protected CrucibleSnapshot createSnapshot() {
            return new CrucibleSnapshot(contained, queued, fluid, renderStack.copy());
        }

        // Compiler dumb
        @Override
        public long extract(ItemVariant resource, long maxAmount, TransactionContext transaction) {
            return InsertionOnlyStorage.super.extract(resource, maxAmount, transaction);
        }

        @Override
        public long getAmount() {
            return 0;
        }

        @Override
        public long getCapacity() {
            return 1;
        }

        @Override
        public ItemVariant getResource() {
            return ItemVariant.blank();
        }

        @Override
        public long insert(ItemVariant resource, long maxAmount, TransactionContext transaction) {
            var recipeOptional = CrucibleRecipe.find(resource.getItem(), isFireproof(), world);
            if (recipeOptional.isEmpty()) return 0;
            var recipe = recipeOptional.get();
            if (!recipe.getFluid().equals(fluid) && !fluid.isBlank()) return 0;

            var amount = Math.min(recipe.getAmount(), getMaxCapacity() - contained - queued);
            if (amount == 0) return 0;
            updateSnapshots(transaction);
            fluid = recipe.getFluid();
            queued += amount;
            renderStack = resource.toStack();
            return 1;
        }

        @Override
        public boolean isResourceBlank() {
            return true;
        }

        @Override
        protected void readSnapshot(CrucibleSnapshot snapshot) {
            contained = snapshot.contained;
            queued = snapshot.queued;
            fluid = snapshot.fluid;
            renderStack = snapshot.renderStack.copy();
        }
    }




}

File: src/main/kotlin/exnihilofabrico/common/crucibles/CrucibleBlockEntity.kt
package exnihilofabrico.common.crucibles

import exnihilofabrico.ExNihiloConfig
import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.ExNihiloRegistries.CRUCIBLE_HEAT
import exnihilofabrico.api.registry.ExNihiloRegistries.CRUCIBLE_STONE
import exnihilofabrico.api.registry.ExNihiloRegistries.CRUCIBLE_WOOD
import exnihilofabrico.common.ModBlocks
import exnihilofabrico.id
import io.github.prospector.silk.fluid.DropletValues
import io.github.prospector.silk.fluid.FluidContainer
import io.github.prospector.silk.fluid.FluidInstance
import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable
import net.minecraft.block.BlockState
import net.minecraft.block.FluidBlock
import net.minecraft.block.Material
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.fluid.Fluid
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.Hand
import net.minecraft.util.Tickable
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.Direction

class CrucibleBlockEntity : BlockEntity(TYPE), FluidContainer, Tickable, BlockEntityClientSerializable {

    var render: ItemStack = ItemStack.EMPTY
    var queued: FluidInstance = FluidInstance.EMPTY
    var contents: FluidInstance = FluidInstance.EMPTY
    var heat: Int = 0

    private var timer: Int = 0
    private val stone: Boolean = world?.getBlockState(pos)?.material == Material.STONE

    override fun markDirty() {
        super.markDirty()

        world?.apply {
            updateListeners(pos, getBlockState(pos), getBlockState(pos), 3)
        }
    }

    override fun tick() {
        if(queued.isEmpty || contents.amount >= maxCapacity || heat <= 0)
            return
        timer += 1
        if(timer > ExNihiloConfig.Modules.Crucibles.tickRate) {
            val amt = Math.min(queued.amount, heat)
            contents.addAmount(amt)
            queued.subtractAmount(amt)
            timer = 0
            markDirty()
        }
    }

    fun activate(state: BlockState?, player: PlayerEntity?, hand: Hand?, hitResult: BlockHitResult?): Boolean {
        if(world?.isClient != false || player == null)
            return true
        val held = player.getStackInHand(hand ?: player.activeHand) ?: ItemStack.EMPTY

        if(held.isEmpty)
            return false

        // Removing a bucket's worth of fluid
        if(held.item == Items.BUCKET) {
            if(contents.amount > DropletValues.BUCKET) {
                val bucket = ItemStack(contents.fluid.bucketItem)
                if(!player.isCreative) {
                    held.subtractAmount(1)
                    contents.subtractAmount(DropletValues.BUCKET)
                }
                player.giveItemStack(bucket)
                markDirty()
            }
            return true
        }
        // Adding a meltable item
        val registry = if(stone) CRUCIBLE_STONE else CRUCIBLE_WOOD
        val result = registry.getResult(held.item)
        if(result != null && // Has result
            (contents.isEmpty || result.fluid == contents.fluid) && // And contents doesn't clash
            (queued.isEmpty || result.fluid == queued.fluid) && // And queued doesn't clash
            queued.amount + contents.amount + result.amount < maxCapacity) {
            render = held.copy()
            if(!player.isCreative)
                held.subtractAmount(1)
            if(queued.isEmpty)
                queued = result.copy()
            else
                queued.addAmount(result.amount)
            markDirty()
            return true
        }
        return true
    }

    fun updateHeat() {
        val oldheat = heat
        val state = world?.getBlockState(pos.down()) ?: return
        val block = state.block
        if(block is FluidBlock) {
            val fluidState = block.getFluidState(state)
            heat = Math.round(CRUCIBLE_HEAT.getHeat(fluidState.fluid) * fluidState.fluid.getHeight(fluidState, world, pos.down())).toInt()
        }
        else {
            heat = CRUCIBLE_HEAT.getHeat(block)
        }
        if(heat != oldheat)
            markDirty()
    }

    /**
     * FluidContainer
     */
    override fun extractFluid(direction: Direction?, fluid: Fluid?, amount: Int) {
        if(contents.fluid == fluid && contents.amount >= amount)
            contents.subtractAmount(amount)
    }

    override fun getMaxCapacity(): Int {
        return 4 * DropletValues.BUCKET
    }

    override fun canExtractFluid(direction: Direction?, fluid: Fluid?, amount: Int): Boolean {
        return contents.fluid == fluid && contents.amount >= amount
    }

    override fun getFluids(direction: Direction?): Array<FluidInstance> {
        return arrayOf(contents)
    }

    override fun setFluid(direction: Direction?, fluid: FluidInstance?) {
        contents = fluid ?: FluidInstance.EMPTY
    }

    override fun canInsertFluid(direction: Direction?, fluid: Fluid?, amount: Int): Boolean {
        return false
    }

    override fun insertFluid(direction: Direction?, fluid: Fluid?, amount: Int) {
        return // Do nothing
    }

    /**
     * NBT Serialization section
     */

    override fun toTag(tag: CompoundTag) = toTagWithoutWorldInfo(super.toTag(tag))

    override fun fromTag(tag: CompoundTag?) {
        super.fromTag(tag)
        if(tag==null){
            ExNihiloFabrico.LOGGER.warn("A crucible at $pos is missing data.")
            return
        }
        fromTagWithoutWorldInfo(tag)
    }

    override fun toClientTag(tag: CompoundTag?) = toTag(tag ?: CompoundTag())
    override fun fromClientTag(tag: CompoundTag?) = fromTag(tag ?: CompoundTag())

    fun toTagWithoutWorldInfo(tag: CompoundTag): CompoundTag {
        tag.put("contents", contents.toTag(CompoundTag()))
        tag.put("queued", queued.toTag(CompoundTag()))
        tag.put("render", render.toTag(CompoundTag()))
        tag.putInt("heat", heat)
        return tag
    }

    fun fromTagWithoutWorldInfo(tag: CompoundTag) {
        contents.fromTag(tag.getCompound("contents"))
        queued.fromTag(tag.getCompound("queued"))
        render = ItemStack.fromTag(tag.getCompound("render"))
        heat = tag.getInt("heat")
    }

    companion object {
        @JvmStatic
        val TYPE: BlockEntityType<CrucibleBlockEntity> =
            BlockEntityType.Builder.create({ CrucibleBlockEntity() }, ModBlocks.CRUCIBLES.values.toTypedArray()).build(null)
        val BLOCK_ENTITY_ID = id("crucible")
    }
}